### ISSUE TITLE
Use NestedTensor in RNN models

### ISSUE THREAD

**Issue by smorad** [pytorch/pytorch#82534]
### 🚀 The feature, motivation and pitch

Now that NestedTensor is in core, I believe a high-impact application would be to replace many RNN utils, such as:

```
torch.nn.utils.rnn.pad_sequence
torch.nn.utils.rnn.pad_packed_sequence
torch.nn.utils.rnn.pack_sequence
```
with nested tensors. LSTM/GRUs are still heavily used in RL, where long episodes can result in a batch of sequences from 1 to 10,000 timesteps. The entire batch must be zero-padded to the length of the longest episode, which uses a ton of memory (or perhaps the RNN utils do something smarter). The RNN utils also pass around list of indices and sort the outputs, which the user must unsort. Furthermore, the above methods are not flexible -- they do not allow a sequence of zero length, for example.

It would be fantastic to stack a bunch of episodes/rollouts into a ragged NestedTensor and pass this straight into an LSTM. I believe all that's needed are sigmoid, tanh, and linear operations which I believe NestedTensor already supports.

**Comment by jbschlosser** [comment]
Hey @smorad, thanks for the feature request! This is definitely something we plan to look into - it's just a matter of when. We have been prioritizing Transformers first, but if you'd like to contribute, I'm happy to review any exploration into NT integration within RNNs.

**Comment by uditrana** [comment]
I think I agree on the prioritization for transformers because those are more bleeding edge but we are seeing more and more use cases where transformers are being combined with LSTMs/RNNs for long sequences. That use case would be really helped with built-in nested tensor support.

**Comment by cpuhrsch** [comment]
@uditrana - We could work on a conversion from NestedTensor to [PackedSequence](https://pytorch.org/docs/stable/generated/torch.nn.utils.rnn.PackedSequence.html)? No promises, but what do you think about the approach?

**Comment by uditrana** [comment]
So right now, the workflow I commonly see (assume a batch first world) is 

1. DataLoader feeds a padded batched input to the forward pass along with the sequence lengths
2. Non-recurrent layers operating on the non-ragged dimension are applied
3. Using padded data and seq lengths, we construct a PackedSequence using pack_padded_sequence
4. Apply RNN layers
5. Use pad_packed_sequence to get a non-ragged padded tensor again (BxTx*)
6. Apply output layers and return

The problem here, putlined by OP as well, is that nestedTensors help us avoid all the unnecessary padding and operations over those padded values. Functionally, it seems they serve a similar purpose to PackedSequence where we don't want to operate over pads in expensive layers like Transformers and RNNs.

So, in a world where we replace the BxTx* padded tensor with a NestedTensor, I think you are right in that all we would need is the equivalents of pack_padded_sequence and pad_packed_sequence taking us from NestedTensors to PackedSequences and back. I think it could be provided as a separate function or just done internally (if we detect the type of the input is nestedTensor). 

I wonder if in the meantime, we could use [torch.nested.to_padded_tensor](https://pytorch.org/docs/stable/nested.html#torch.nested.to_padded_tensor) as an expensive but functional stopgap (NestedTensor -> PaddedTensor -> PackedSequence).

**Comment by cpuhrsch** [comment]
@uditrana - Yes that makes sense. The layout used by PackedSequence is a big peculiar so there will be a conversion step necessary in any case. Adding support for NestedTensor to [pack_padded_sequence](https://pytorch.org/docs/stable/generated/torch.nn.utils.rnn.pack_sequence.html#torch.nn.utils.rnn.pack_sequence) could be interesting. We don't really have the time to work on this right now, but are happy to review a community contribution if you're interested?

In particular we could add a `torch.nested.to_packed_sequence` and `torch.nested.from_packed_sequence`. A first step could be using the `to_padded_tensor` and private [_nested_from_padded](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L5466) as implementation detail.

**Comment by uditrana** [comment]
Hi again,

So in the last 2 weeks, it seems I was able to implement the workflow I described above without any additional functionality. Basically to implement the workflow above we need a tensor object that can be batched despite having one dimension (for sequence length) as ragged (shape BxSx*).
This object needs to support 2 types of functionality:
1. Layers and functions over the non-sequence dimensions without padding
2. recurrent and transformer operations which do operate over the ragged sequence dimension.

PackedSequence supports #2 but not #1, while NestedTensor currently supports #1 but not #2. 

That said, I found issues [#25310](https://github.com/pytorch/pytorch/issues/25310) and [#59629](https://github.com/pytorch/pytorch/issues/59629) which discuss the implementation of `squash_packed`. 
```
def squash_packed(x, fn=torch.tanh):
    return torch.nn.utils.rnn.PackedSequence(fn(x.data), x.batch_sizes, 
                 x.sorted_indices, x.unsorted_indices)
```

This function basically applies any generic pytorch function over the  * dimension while collapsing Batch and Ragged Sequence dimensions together. So even though the code is ugly, this has the ability to basically give PackedSequence the desired capability #1 it was missing. And as far as I can tell, I haven't seen major performance hits.

So long story short, I was able to implement the workflow without needing NestedTensors _for now_. I think it is possible when I get time, I will look refactoring our models with NestedTensors. When that happens, I will probably take a deeper dive to understand how NestedTensors work and see if I can build a helper that implements #2 above for nested tensors. If it seems useful, I could try making that into a community contribution.

But it's unclear how much I'd gain outside of code clarity from switching. Do you have thoughts or something I'm missing?

**Comment by cpuhrsch** [comment]
@uditrana - That seems reasonable. cc @drisspg to track RNN integration.
